Return the best (lowest energy) state instead of the final state

paceheart opened the pull request:
The final state isn't necessarily the best (lowest-energy) state. Now Simulator::run keeps track of the best state as it runs, and returns the best state instead of the happenstance final state.

paceheart committed:
refactor to expose lower_energy? method

paceheart committed:
keep track of the best (lowest energy) state and return it

The final state isn't necessarily the best (lowest-energy) state. Now Simulator::run keeps track of the best state as it runs, and returns the best state instead of the happenstance final state.

paceheart committed:
revert accidental comment edit

paceheart committed:
revert accidental comment edit

paceheart commented:
Oops! Sorry for not running the tests first. I'll take a look.

paceheart committed:
remove whitespace

paceheart committed:
preserve the final temperature in the return value

also factor out a helper method, to get out of ruby jail

paceheart committed:
raise a RuntimeError if the solution energy is greater than the initial energy

rake just displays the error and moves merrily along, but that's better than nothing

paceheart committed:
add test_finds_optimal_solution

This test (almost always) fails before the "keep track of the best state" patch, and consistently succeeds after.

paceheart commented:
fixed!

3zcurdia commented on lib/annealing/metal.rb:
```diff
@@ -27,14 +27,21 @@ def cool!(new_temperature)
       end
     end
 
+    # True if cooled_metal.energy is lower than current energy.
```
Unnecessary comment, the function is self explanatory 
```suggestion
```

3zcurdia replied:
Also if you create a new function, it should be bellow the prefer? function so it turns easy to read

3zcurdia commented on lib/annealing/metal.rb:
```diff
@@ -27,14 +27,21 @@ def cool!(new_temperature)
     # value
     def prefer?(cooled_metal)
-      return true if cooled_metal.energy < energy
+      lower_energy?(cooled_metal) or prefer_despite_higher_energy?(cooled_metal)
```
In ruby using `||` is preferred over `or` 
```suggestion
      lower_energy?(cooled_metal) || prefer_despite_higher_energy?(cooled_metal)
```

3zcurdia commented on lib/annealing/simulator.rb:
```diff
@@ -13,15 +13,24 @@ def run(initial_state, config_hash = {})
+      until termination_condition_met?(current, runtime_config)
+        steps += 1
+        current = reduce_temperature(current, steps, runtime_config)
+        # If the current state has lower energy than the previous best (lowest energy) state
+        # we've seen so far, the current state is the new best state.
```
```suggestion
```

3zcurdia commented on bin/run:
```diff
@@ -44,14 +44,20 @@ solution = simulator.run(locations,
 end
 puts "-------\nEnergy: #{solution.energy}"
+
+unless solution.energy <= initial_energy
```
Direct logic is always easy to read
```suggestion
if solution.energy > initial_energy
```

3zcurdia reviewed (changes requested):
Thanks for your contribution, but since this gem is being used in exiting projects we need to consider to this change to be available through configuration

3zcurdia commented on lib/annealing/simulator.rb:
```diff
@@ -13,15 +13,24 @@ def run(initial_state, config_hash = {})
-          current = reduce_temperature(current, steps, runtime_config)
-        end
-        current
+        really_run(current, runtime_config)
```
Also extracting this into an smaller method just to make happy rubocop is unnecessary we can tell rubocop to skip the cop on this method

paceheart replied:
thanks! I haven't used rubocop before; I appreciate your patience with me

paceheart committed:
remove unnecessary comment; the function is self explanatory

Co-authored-by: Luis Ezcurdia <ing.ezcurdia@gmail.com>

paceheart committed:
use || instead of 'or'

Co-authored-by: Luis Ezcurdia <ing.ezcurdia@gmail.com>

paceheart committed:
improve readability

Co-authored-by: Luis Ezcurdia <ing.ezcurdia@gmail.com>

paceheart committed:
merge really_run back into run

"extracting this into a smaller method just to make rubocop happy is unnecessary; we can tell rubocop to skip the cop on this method" -3zcurdia

paceheart committed:
Merge branch 'main' of https://github.com/paceheart/annealing

paceheart committed:
Add return_best configuration option

When true, returns the best (lowest energy) state. When false, returns the final state (preserves old behavior).

paceheart commented:
> Thanks for your contribution, but since this gem is being used in exiting projects we need to consider to this change to be available through configuration

I added a new configuration option 'return_best'. I set the default to true, but if you think it's more important to maintain backward compatibility, we can set it to false instead.

chrisbloom7 commented on test/annealing/simulator_test.rb:
```diff
@@ -123,5 +123,44 @@ def test_passes_run_time_configuration_to_metal
+      assert_equal 1, final_energy
+    end
+
+    def test_finds_suboptimal_solution_when_not_return_best
+      max_iterations = 10
+      initial_energy = real_energy_calculator(@collection)
+      final_energy = 1
+      iteration_count = 0
+
+      # The simulator might happen to finish on the optimal state by random chance.
+      # If that happens, run it again up to MAX_ITERATIONS times.
+      until final_energy > 1 || iteration_count > max_iterations
+        iteration_count += 1
+        final_metal = @simulator.run(@collection,
+                                     energy_calculator: ->(x) { real_energy_calculator(x) },
+                                     state_change: ->(state) { state.shuffle },
+                                     return_best: false)
+        final_energy = real_energy_calculator(final_metal.state)
+
+        assert final_energy < initial_energy
+      end
+
+      refute_equal 1, final_energy
+    end
```
There's still a chance this test could fail after max_attempts since it's just relying on random chance. I'd prefer a test that is deterministic, even if that means using stubs or mocks to setup the specific scenario this test is exercising.

3zcurdia replied:
I think I could take it from here once I merge it

chrisbloom7 commented on lib/annealing/configuration.rb:
```diff
@@ -5,12 +5,14 @@ module Annealing
   class Configuration
     DEFAULT_COOLING_RATE = 0.0003
     DEFAULT_INITIAL_TEMPERATURE = 10_000.0
+    DEFAULT_RETURN_BEST = true
```
I think that this should default to `false` if this will end up going into a minor release. If we want to make it default to `true`, then we should def give a major version bump and I'd expect to see documentation about the upgrade path to keep the old functionality. I defer to @3zcurdia about which route he would prefer to see.

3zcurdia replied:
I totally agree  we must keep compatibility with the current behaviour to keep as false
```suggestion
    DEFAULT_RETURN_BEST = false
```

chrisbloom7 reviewed (changes requested):
Good suggestion, thanks for the contribution. Left some feedback below.

3zcurdia committed:
Set default as false
